File: frontend/src/api/review.ts
/**
 * 交易复盘 API
 */
import { ApiClient } from './request'

// ==================== 类型定义 ====================

/** 复盘类型 */
export type ReviewType = 'single_trade' | 'complete_trade' | 'periodic'

/** 复盘状态 */
export type ReviewStatus = 'pending' | 'processing' | 'completed' | 'failed'

/** 交易方向 */
export type TradeSide = 'buy' | 'sell'

/** 交易记录 */
export interface TradeRecord {
  trade_id: string
  code: string
  market: string
  side: TradeSide
  quantity: number
  price: number
  amount: number
  pnl: number
  commission: number
  timestamp: string
  change_type?: string
  quantity_after?: number
  created_at?: string
}

/** 交易信息汇总 */
export interface TradeInfo {
  code: string
  name?: string
  market: string
  trades: TradeRecord[]
  total_buy_quantity: number
  total_sell_quantity: number
  avg_buy_price: number
  avg_sell_price: number
  total_buy_amount: number
  total_sell_amount: number
  total_commission: number
  realized_pnl: number
  realized_pnl_pct: number
  unrealized_pnl?: number  // 🆕 浮动盈亏（持仓中时）
  unrealized_pnl_pct?: number  // 🆕 浮动盈亏百分比（持仓中时）
  current_price?: number  // 🆕 当前价格（持仓中时）
  is_holding?: boolean  // 🆕 是否还在持仓中
  remaining_quantity?: number  // 🆕 剩余持仓数量
  first_buy_date?: string
  last_sell_date?: string
  holding_days: number
}

/** 市场数据快照 */
export interface MarketSnapshot {
  period_high?: number
  period_high_date?: string
  period_low?: number
  period_low_date?: string
  buy_date_open?: number
  buy_date_high?: number
  buy_date_low?: number
  buy_date_close?: number
  sell_date_open?: number
  sell_date_high?: number
  sell_date_low?: number
  sell_date_close?: number
  optimal_buy_price?: number
  optimal_buy_date?: string
  optimal_sell_price?: number
  optimal_sell_date?: string
  kline_data?: Array<{
    date: string
    open: number
    high: number
    low: number
    close: number
    volume: number
  }>
}

/** AI复盘分析结果 */
export interface AITradeReview {
  overall_score: number
  timing_score?: number
  position_score?: number
  discipline_score?: number
  emotion_score?: number
  attribution_score?: number
  summary: string
  strengths: string[]
  weaknesses: string[]
  suggestions: string[]  // 兼容字段名，实际展示为复盘改进项
  timing_analysis?: string
  position_analysis?: string
  emotion_analysis?: string
  attribution_analysis?: string
  actual_pnl?: number
  optimal_pnl?: number
  missed_profit?: number
  avoided_loss?: number
  plan_adherence?: string  // 计划执行情况
  plan_deviation?: string   // 偏离说明
}

/** 复盘报告 */
export interface TradeReviewReport {
  review_id: string
  task_id?: string
  review_type: ReviewType
  status: ReviewStatus
  source?: 'paper' | 'position' | 'real'
  error_message?: string
  trade_info: TradeInfo
  market_snapshot: MarketSnapshot
  ai_review: AITradeReview
  is_case_study: boolean
  tags: string[]
  execution_time?: number
  created_at: string
  trading_system_id?: string  // 关联的交易计划ID
  trading_system_name?: string  // 关联的交易计划名称
}

/** 复盘列表项 */
export interface ReviewListItem {
  review_id: string
  review_type: ReviewType
  code?: string
  name?: string
  realized_pnl: number
  overall_score: number
  status: ReviewStatus
  is_case_study: boolean
  created_at: string
}

/** 交易统计 */
export interface TradingStatistics {
  total_trades: number
  winning_trades: number
  losing_trades: number
  win_rate: number
  total_pnl: number
  avg_profit: number
  avg_loss: number
  profit_loss_ratio: number
  max_single_profit: number
  max_single_loss: number
  total_commission: number
}

/** 可复盘的股票 */
export interface ReviewableStock {
  code: string
  name?: string
  market?: string
  status?: 'completed' | 'holding'  // completed=已完成交易, holding=持仓中
  buy_count: number
  sell_count: number
  total_pnl: number
  total_buy_amount?: number
  total_sell_amount?: number
  last_trade_time?: string
}

// ==================== 请求类型 ====================

/** 创建复盘请求 */
export interface CreateTradeReviewRequest {
  trade_ids: string[]
  review_type?: ReviewType
  code?: string
  source?: 'real' | 'paper'  // 数据源：用户持仓或模拟持仓
  trading_system_id?: string  // 关联的交易计划ID
  workflow_id?: string  // 关联的工作流ID
  use_workflow?: boolean  // 是否使用工作流引擎（v2.0）
}

/** 保存案例请求 */
export interface SaveAsCaseRequest {
  review_id: string
  tags?: string[]
  source?: 'paper' | 'position' | 'real'  // 数据源: paper(模拟交易) 或 position(持仓操作)
}

/** 复盘历史筛选参数 */
export interface ReviewHistoryFilter {
  page?: number
  pageSize?: number
  code?: string
  startDate?: string
  endDate?: string
  reviewType?: ReviewType
  source?: 'paper' | 'position'  // 数据源: paper(模拟交易) 或 position(持仓操作)
}

/** 可复盘交易筛选参数 */
export interface ReviewableTradesFilter {
  code?: string
  startDate?: string
  endDate?: string
  page?: number
  pageSize?: number
}

/** 阶段性复盘周期类型 */
export type PeriodType = 'week' | 'month' | 'quarter' | 'year'

/** 创建阶段性复盘请求 */
export interface CreatePeriodicReviewRequest {
  period_type: PeriodType
  start_date: string
  end_date: string
  source?: 'paper' | 'position'  // 数据源: paper(模拟交易) 或 position(持仓操作)
}

/** AI阶段性复盘结果 */
export interface AIPeriodicReview {
  overall_score: number
  summary: string
  trading_style: string
  common_mistakes: string[]
  improvement_areas: string[]  // 复盘改进方向
  action_plan: string[]  // 兼容字段名，实际展示为后续观察项
  best_trade?: string
  worst_trade?: string
}

/** 交易摘要项 */
export interface TradeSummaryItem {
  code: string
  name?: string
  side: string
  quantity: number
  price: number
  pnl: number
  pnl_pct: number
  timestamp: string
}

/** 阶段性复盘报告 */
export interface PeriodicReviewReport {
  review_id: string
  period_type: PeriodType
  period_start: string
  period_end: string
  statistics: TradingStatistics
  trades_summary: TradeSummaryItem[]
  ai_review: AIPeriodicReview
  status: ReviewStatus
  execution_time?: number
  created_at: string
}

/** 阶段性复盘列表项 */
export interface PeriodicReviewListItem {
  review_id: string
  period_type: PeriodType
  period_start: string
  period_end: string
  source?: 'paper' | 'position'
  total_trades: number
  total_pnl: number
  win_rate: number
  overall_score: number
  status: ReviewStatus
  created_at: string
}

// ==================== API 方法 ====================

export const reviewApi = {
  /** 创建交易复盘 */
  async createTradeReview(data: CreateTradeReviewRequest) {
    return ApiClient.post<TradeReviewReport>('/api/review/trade', data, {
      showLoading: true,
      timeout: 30000,  // 30秒超时（后端立即返回任务ID，不需要等待任务完成）
      retryCount: 0     // 禁用重试，避免重复分析
    })
  },

  /** 获取复盘历史列表，支持筛选 */
  async getReviewHistory(filter: ReviewHistoryFilter = {}) {
    const { page = 1, pageSize = 10, code, startDate, endDate, reviewType, source } = filter
    return ApiClient.get<{ items: ReviewListItem[]; total: number; page: number; page_size: number }>(
      '/api/review/trade/history',
      {
        page,
        page_size: pageSize,
        code,
        start_date: startDate,
        end_date: endDate,
        review_type: reviewType,
        source
      }
    )
  },

  /** 获取复盘详情 */
  async getReviewDetail(reviewId: string) {
    return ApiClient.get<TradeReviewReport>(`/api/review/trade/${reviewId}`)
  },

  /** 保存为案例 */
  async saveAsCase(data: SaveAsCaseRequest) {
    return ApiClient.post<{ message: string }>('/api/review/case', data, { showLoading: true })
  },

  /** 获取案例库 */
  async getCases(params: { page?: number; pageSize?: number; source?: 'paper' | 'position' } = {}) {
    const { page = 1, pageSize = 10, source } = params
    const queryParams: Record<string, any> = { page, page_size: pageSize }
    if (source) {
      queryParams.source = source
    }
    return ApiClient.get<{ items: ReviewListItem[]; total: number; page: number; page_size: number }>(
      '/api/review/cases',
      queryParams
    )
  },

  /** 从案例库删除 */
  async deleteCase(reviewId: string) {
    return ApiClient.delete<{ message: string }>(`/api/review/case/${reviewId}`, { showLoading: true })
  },

  /** 删除复盘记录 */
  async deleteReview(reviewId: string) {
    return ApiClient.delete<{ message: string }>(`/api/review/trade/${reviewId}`, { showLoading: true })
  },

  /** 获取交易统计 */
  async getTradingStatistics(startDate?: string, endDate?: string) {
    return ApiClient.get<TradingStatistics>('/api/review/statistics', {
      start_date: startDate,
      end_date: endDate
    })
  },

  /** 获取可复盘的交易列表，支持筛选 */
  async getReviewableTrades(filter: ReviewableTradesFilter = {}) {
    const { code, startDate, endDate, page = 1, pageSize = 20 } = filter
    return ApiClient.get<{
      items: TradeRecord[]
      total: number
      page: number
      page_size: number
      completed_stocks: ReviewableStock[]
      all_stocks: ReviewableStock[]  // 所有有交易的股票（包括只买入的）
    }>('/api/review/reviewable-trades', {
      code,
      start_date: startDate,
      end_date: endDate,
      page,
      page_size: pageSize
    })
  },

  /** 获取某只股票的所有交易 */
  async getTradesByCode(code: string, source: 'real' | 'paper' = 'real') {
    return ApiClient.get<{
      code: string
      trades: TradeRecord[]
      summary: {
        total_trades: number
        total_buy_quantity: number
        total_sell_quantity: number
        total_pnl: number
        is_closed: boolean
      }
    }>(`/api/review/trades-by-code/${code}`, { source })
  },

  // ==================== 阶段性复盘 ====================

  /** 创建阶段性复盘 */
  async createPeriodicReview(data: CreatePeriodicReviewRequest) {
    return ApiClient.post<PeriodicReviewReport>('/api/review/periodic', data, {
      showLoading: true,
      timeout: 600000,  // 600秒超时（10分钟）
      retryCount: 0     // 禁用重试，避免重复分析
    })
  },

  /** 获取阶段性复盘历史 */
  async getPeriodicReviewHistory(page = 1, pageSize = 10, source?: 'paper' | 'position') {
    const params: Record<string, any> = { page, page_size: pageSize }
    if (source) {
      params.source = source
    }
    return ApiClient.get<{ items: PeriodicReviewListItem[]; total: number; page: number; page_size: number }>(
      '/api/review/periodic/history',
      params
    )
  },

  /** 获取阶段性复盘详情 */
  async getPeriodicReviewDetail(reviewId: string) {
    return ApiClient.get<PeriodicReviewReport>(`/api/review/periodic/${reviewId}`)
  },

  /** 按股票代码获取复盘摘要（用于股票详情页） */
  async getTradeReviewSummary(code: string) {
    return ApiClient.get<ReviewSummary>(`/api/review/summary`, { code })
  }
}

/** 复盘摘要类型（股票详情页使用） */
interface ReviewSummary {
  has: boolean
  task_id?: string
  completed_at?: string
  summary?: string
  recommendation?: string
  risk_level?: string
  confidence_score?: number | null
  trade_count?: number | null
}

File: app/models/review.py
"""
交易复盘数据模型
"""

from datetime import datetime
from typing import Optional, List
from enum import Enum
from pydantic import BaseModel, Field, ConfigDict
from bson import ObjectId

from app.utils.timezone import now_tz


class PyObjectId(ObjectId):
    @classmethod
    def __get_validators__(cls):
        yield cls.validate

    @classmethod
    def validate(cls, v, info=None):
        if not ObjectId.is_valid(v):
            raise ValueError("Invalid ObjectId")
        return ObjectId(v)

    @classmethod
    def __get_pydantic_json_schema__(cls, schema, handler):
        return {"type": "string"}


class ReviewType(str, Enum):
    """复盘类型"""
    SINGLE_TRADE = "single_trade"      # 单笔交易复盘
    COMPLETE_TRADE = "complete_trade"  # 完整交易复盘（买入到卖出）
    PERIODIC = "periodic"              # 阶段性复盘


class PeriodType(str, Enum):
    """阶段性复盘周期类型"""
    WEEK = "week"
    MONTH = "month"
    QUARTER = "quarter"
    YEAR = "year"


class ReviewStatus(str, Enum):
    """复盘状态"""
    PENDING = "pending"
    PROCESSING = "processing"
    COMPLETED = "completed"
    FAILED = "failed"


# ==================== 交易信息相关模型 ====================

class TradeRecord(BaseModel):
    """交易记录"""
    trade_id: Optional[str] = None
    side: str                           # buy/sell
    quantity: int
    price: float
    amount: float = 0.0
    commission: float = 0.0
    timestamp: str
    pnl: float = 0.0                    # 实现盈亏（卖出时）


class TradeInfo(BaseModel):
    """交易信息汇总"""
    code: str
    name: Optional[str] = None
    market: str = "CN"
    currency: str = "CNY"
    trades: List[TradeRecord] = Field(default_factory=list)
    
    # 汇总数据
    total_buy_quantity: int = 0
    total_buy_amount: float = 0.0
    avg_buy_price: float = 0.0
    total_sell_quantity: int = 0
    total_sell_amount: float = 0.0
    avg_sell_price: float = 0.0
    
    # 盈亏
    realized_pnl: float = 0.0           # 已实现盈亏（已卖出部分）
    realized_pnl_pct: float = 0.0       # 已实现盈亏百分比
    unrealized_pnl: float = 0.0         # 🆕 未实现盈亏（持仓部分）
    unrealized_pnl_pct: float = 0.0     # 🆕 未实现盈亏百分比
    total_commission: float = 0.0

    # 持仓状态
    is_holding: bool = False            # 🆕 是否还在持仓中
    current_price: Optional[float] = None  # 🆕 当前价格（持仓中时）
    remaining_quantity: int = 0         # 🆕 剩余持仓数量

    # 持仓周期
    first_buy_date: Optional[str] = None
    last_sell_date: Optional[str] = None
    holding_days: int = 0


class MarketSnapshot(BaseModel):
    """市场数据快照"""
    # 买入时行情
    buy_date_open: Optional[float] = None
    buy_date_high: Optional[float] = None
    buy_date_low: Optional[float] = None
    buy_date_close: Optional[float] = None
    
    # 卖出时行情
    sell_date_open: Optional[float] = None
    sell_date_high: Optional[float] = None
    sell_date_low: Optional[float] = None
    sell_date_close: Optional[float] = None
    
    # 持仓期间行情
    period_high: Optional[float] = None
    period_high_date: Optional[str] = None
    period_low: Optional[float] = None
    period_low_date: Optional[str] = None
    
    # 最优买卖点
    optimal_buy_price: Optional[float] = None
    optimal_sell_price: Optional[float] = None
    
    # K线数据（用于图表展示）
    kline_data: List[dict] = Field(default_factory=list)


# ==================== AI分析结果模型 ====================

class AITradeReview(BaseModel):
    """AI交易复盘结果

    评分体系说明：
    - overall_score: 综合评分 = (timing_score + position_score + emotion_score + attribution_score) / 4
    - timing_score: 时机评分（买卖时机选择）
    - position_score: 仓位评分（仓位管理）
    - emotion_score: 情绪与纪律评分（情绪控制 + 纪律执行）
    - attribution_score: 归因评分（收益来源分析）
    - discipline_score: 已废弃，保留仅为向后兼容（纪律评估已合并到 emotion_score）
    """
    overall_score: int = Field(0, ge=0, le=100, description="总评分")
    timing_score: int = Field(0, ge=0, le=100, description="时机评分")
    position_score: int = Field(0, ge=0, le=100, description="仓位评分")
    emotion_score: int = Field(0, ge=0, le=100, description="情绪与纪律评分（情绪控制50% + 纪律执行50%）")
    attribution_score: int = Field(0, ge=0, le=100, description="归因评分")
    discipline_score: int = Field(0, ge=0, le=100, description="[已废弃] 纪律评分（已合并到emotion_score，保留仅为兼容性）")

    summary: str = ""                                # 总结
    strengths: List[str] = Field(default_factory=list)  # 做得好的地方
    weaknesses: List[str] = Field(default_factory=list) # 需要改进的地方
    suggestions: List[str] = Field(default_factory=list) # 兼容旧字段名，实际语义为复盘改进项

    timing_analysis: str = ""                        # 时机分析详情
    position_analysis: str = ""                      # 仓位分析详情
    emotion_analysis: str = ""                       # 情绪分析详情
    attribution_analysis: str = ""                   # 归因分析详情

    # 收益对比
    actual_pnl: float = 0.0                          # 实际收益
    optimal_pnl: float = 0.0                         # 理论最优收益
    missed_profit: float = 0.0                       # 错失的收益
    avoided_loss: float = 0.0                        # 避免的亏损

    # 🆕 交易计划执行情况（当关联了交易计划时）
    plan_adherence: Optional[str] = None             # 计划执行良好的地方
    plan_deviation: Optional[str] = None             # 偏离计划的地方


# ==================== 复盘报告模型 ====================

class TradeReviewReport(BaseModel):
    """交易复盘报告 (trade_reviews 集合)"""
    id: Optional[PyObjectId] = Field(default_factory=PyObjectId, alias="_id")
    review_id: str
    user_id: str
    review_type: ReviewType = ReviewType.COMPLETE_TRADE
    
    # 交易信息
    trade_info: TradeInfo = Field(default_factory=TradeInfo)
    
    # 市场数据快照
    market_snapshot: MarketSnapshot = Field(default_factory=MarketSnapshot)
    
    # AI分析结果
    ai_review: AITradeReview = Field(default_factory=AITradeReview)
    
    # 状态与元数据
    status: ReviewStatus = ReviewStatus.PENDING
    error_message: Optional[str] = None
    execution_time: float = 0.0
    
    # 案例相关
    is_case_study: bool = False
    tags: List[str] = Field(default_factory=list)
    source: str = Field(default="paper", description="数据源: paper(模拟交易) 或 position(持仓操作)")

    # 交易计划关联
    trading_system_id: Optional[str] = Field(None, description="关联的交易计划ID")
    trading_system_name: Optional[str] = Field(None, description="关联的交易计划名称")

    created_at: datetime = Field(default_factory=now_tz)
    
    model_config = ConfigDict(
        populate_by_name=True,
        arbitrary_types_allowed=True,
        json_encoders={ObjectId: str, datetime: lambda v: v.isoformat()}
    )


# ==================== 阶段性复盘模型 ====================

class TradingStatistics(BaseModel):
    """交易统计"""
    total_trades: int = 0                # 总交易次数
    winning_trades: int = 0              # 盈利次数
    losing_trades: int = 0               # 亏损次数
    win_rate: float = 0.0                # 胜率

    total_pnl: float = 0.0               # 总盈亏
    avg_profit: float = 0.0              # 平均盈利
    avg_loss: float = 0.0                # 平均亏损
    profit_loss_ratio: float = 0.0       # 盈亏比

    max_single_profit: float = 0.0       # 最大单笔盈利
    max_single_loss: float = 0.0         # 最大单笔亏损
    max_drawdown: float = 0.0            # 最大回撤

    avg_holding_days: float = 0.0        # 平均持仓天数
    total_commission: float = 0.0        # 总手续费


class AIPeriodicReview(BaseModel):
    """AI阶段性复盘结果"""
    overall_score: int = Field(0, ge=0, le=100)
    summary: str = ""

    trading_style: str = ""              # 交易风格分析
    common_mistakes: List[str] = Field(default_factory=list)   # 常见错误
    improvement_areas: List[str] = Field(default_factory=list) # 复盘改进方向
    action_plan: List[str] = Field(default_factory=list)       # 兼容旧字段名，实际语义为后续观察项

    best_trade: Optional[str] = None     # 最佳交易分析
    worst_trade: Optional[str] = None    # 最差交易分析


class TradeSummaryItem(BaseModel):
    """交易摘要项"""
    code: str
    name: Optional[str] = None
    side: str
    quantity: int
    price: float
    pnl: float = 0.0
    pnl_pct: float = 0.0
    timestamp: str


class PeriodicReviewReport(BaseModel):
    """阶段性复盘报告 (periodic_reviews 集合)"""
    id: Optional[PyObjectId] = Field(default_factory=PyObjectId, alias="_id")
    review_id: str
    user_id: str

    # 复盘周期
    period_type: PeriodType = PeriodType.MONTH
    period_start: datetime
    period_end: datetime

    # 数据源: paper(模拟交易) 或 position(持仓操作)
    source: str = "paper"

    # 交易统计
    statistics: TradingStatistics = Field(default_factory=TradingStatistics)

    # 交易明细摘要
    trades_summary: List[TradeSummaryItem] = Field(default_factory=list)

    # AI分析
    ai_review: AIPeriodicReview = Field(default_factory=AIPeriodicReview)

    # 状态与元数据
    status: ReviewStatus = ReviewStatus.PENDING
    error_message: Optional[str] = None
    execution_time: float = 0.0

    created_at: datetime = Field(default_factory=now_tz)

    model_config = ConfigDict(
        populate_by_name=True,
        arbitrary_types_allowed=True,
        json_encoders={ObjectId: str, datetime: lambda v: v.isoformat()}
    )


# ==================== API请求/响应模型 ====================

class CreateTradeReviewRequest(BaseModel):
    """创建交易复盘请求"""
    trade_ids: List[str] = Field(..., min_length=1, description="要复盘的交易ID列表")
    review_type: ReviewType = ReviewType.COMPLETE_TRADE
    code: Optional[str] = Field(None, description="股票代码（如果不提供则从交易记录推断）")
    source: str = Field("paper", description="数据源: paper(模拟交易) 或 position(持仓操作)")
    trading_system_id: Optional[str] = Field(None, description="关联的交易计划ID，如果提供则按照交易计划规则进行检查")
    use_workflow: Optional[bool] = Field(None, description="是否使用工作流引擎（v2.0）。None表示使用环境变量设置")


class CreatePeriodicReviewRequest(BaseModel):
    """创建阶段性复盘请求"""
    period_type: PeriodType = PeriodType.MONTH
    start_date: str = Field(..., description="开始日期 YYYY-MM-DD")
    end_date: str = Field(..., description="结束日期 YYYY-MM-DD")
    source: str = Field("paper", description="数据源: paper(模拟交易) 或 position(持仓操作)")


class SaveAsCaseRequest(BaseModel):
    """保存为案例请求"""
    review_id: str
    tags: List[str] = Field(default_factory=list)
    notes: Optional[str] = None
    source: Optional[str] = Field(None, description="数据源: paper(模拟交易) 或 position(持仓操作)，如果不提供则从复盘报告中读取")


class TradeReviewResponse(BaseModel):
    """交易复盘响应"""
    review_id: str
    status: ReviewStatus
    trade_info: Optional[TradeInfo] = None
    ai_review: Optional[AITradeReview] = None
    market_snapshot: Optional[MarketSnapshot] = None
    execution_time: float = 0.0
    created_at: Optional[datetime] = None


class PeriodicReviewResponse(BaseModel):
    """阶段性复盘响应"""
    review_id: str
    status: ReviewStatus
    period_type: PeriodType
    period_start: Optional[datetime] = None
    period_end: Optional[datetime] = None
    statistics: Optional[TradingStatistics] = None
    ai_review: Optional[AIPeriodicReview] = None
    execution_time: float = 0.0
    created_at: Optional[datetime] = None


class ReviewListItem(BaseModel):
    """复盘列表项"""
    review_id: str
    review_type: str
    code: Optional[str] = None
    name: Optional[str] = None
    realized_pnl: float = 0.0
    overall_score: int = 0
    status: str
    is_case_study: bool = False
    created_at: Optional[datetime] = None

